### PR TITLE
scheleton for adding stepchain parentage

### PR DESCRIPTION
**Pull request by ticoann**
replace #8676 
#8700 need to be merged first

### CODE REVIEW

**Comment by ticoann** [comment]
@amaltaro, Alan as we discussed I change to get the parentage information from spec instead of wmstats. It is a little more intrusive and complicated. Could you review the logic? I still need to test.

**Comment by ticoann** [comment]
It assumes WMSpec helper has getParentDataset(dataset). I will add that function when your patch is in.

File: src/python/WMComponent/DBS3Buffer/DBSBufferBlock.py [hunk 1]
```diff
@@ -211,14 +215,14 @@ def addBlockParent(self, parent):
         self.data['block_parent_list'].append({'block_name': parent})
         return
 
-    def addDatasetParent(self, parent):
+    def addDatasetParents(self, parents):
```
Comment (amaltaro) [inline src/python/WMComponent/DBS3Buffer/DBSBufferBlock.py]: Why is it a list if there is only one parent dataset? I'd suggest to call it singular too

File: src/python/WMComponent/DBS3Buffer/DBSBufferDataset.py [hunk 2]
```diff
@@ -17,9 +17,8 @@ class DBSBufferDataset(WMConnectionBase, dict):
     information
     """
 
-    def __init__(self, path, id = -1,
-                 processingVer = None, acquisitionEra = None, validStatus = None,
-                 globalTag = None, parent = None, prep_id = None):
+    def __init__(self, path, id=-1,processingVer=None, acquisitionEra=None, validStatus=None,
+                 globalTag=None, parent=None, prep_id=None, stepChain=0):
```
Comment (amaltaro) [inline src/python/WMComponent/DBS3Buffer/DBSBufferDataset.py]: Minor, but will be helpful in the future :) How about renaming it to `is_stepchain`?

File: src/python/WMComponent/DBS3Buffer/DBSBufferBlock.py (hunk 1 above)
Comment (amaltaro) [inline src/python/WMComponent/DBS3Buffer/DBSBufferBlock.py]: Why is it a list if there is only one parent dataset? I'd suggest to change it to singular too

File: src/python/WMComponent/DBS3Buffer/DBSBufferBlock.py [hunk 3]
```diff
@@ -152,9 +152,13 @@ def addFile(self, dbsFile, datasetType, primaryDatasetType):
         self.data['files'].append(fileDict)
 
         # now add file to data
-        parentLFNs = dbsFile.getParentLFNs()
-        for lfn in parentLFNs:
-            self.addFileParent(child = dbsFile['lfn'], parent = lfn)
+        # If dataset_parent_list is defined don't add the file parentage.
+        # This means it is block from StepChain workflow and parentage of file will be resloved later
```
Comment (amaltaro) [inline src/python/WMComponent/DBS3Buffer/DBSBufferBlock.py]: typo and I think this blank line can go away. Or you fix it in a later commit.

File: src/python/WMComponent/DBS3Buffer/DBSUploadPoller.py [hunk 4]
```diff
@@ -341,6 +341,10 @@ def loadBlocks(self):
             block = DBSBufferBlock(name=blockInfo['block_name'],
                                    location=blockInfo['origin_site_name'],
                                    datasetpath=blockInfo['datasetpath'])
+
+            parents = self.getStepChainParentage(blockInfo['datasetpath'])
+            block.addDatasetParents(parents)
```
Comment (amaltaro) [inline src/python/WMComponent/DBS3Buffer/DBSUploadPoller.py]: `addParentDataset`?

File: src/python/WMComponent/DBS3Buffer/DBSUploadPoller.py [hunk 5]
```diff
@@ -816,3 +822,16 @@ def checkBlocks(self):
 
         # We're done
         return
+
+    def getStepChainParentage(self, dataset):
+        myThread = threading.currentThread()
+        parentDatasetDAO = self.daoFactory(classname="UpdateBlocks")
```
Comment (amaltaro) [inline src/python/WMComponent/DBS3Buffer/DBSUploadPoller.py]: How about defining this DAO (and actually all the others) inside the __init__ method? And is it really going to be an update operation or just a select? Don't forget you need to create a new DAO for this.

In addition to that - there is no need to fix it in this PR - I just noticed we don't have the transaction handle in the algorithm method. In the last case it's handled in the BaseWorkerThread or something like that, but shouldn't we have the rollback and etc in algorithm?

Comment (amaltaro) [inline src/python/WMComponent/DBS3Buffer/DBSUploadPoller.py]: Haa, you probably meant to write `GetStepChainParentDataset` :)

File: src/python/WMComponent/DBS3Buffer/DBSUploadPoller.py (hunk 4 above)
Comment (amaltaro) [reply src/python/WMComponent/DBS3Buffer/DBSUploadPoller.py]: And when it's not a StepChain block, the DAO will return an empty list or something like that, shouldn't we then skip this `block.addDatasetParents` call such that we leave the correct parentage information for non-stepchain requests?

File: src/python/WMCore/WorkQueue/WMBSHelper.py [hunk 6]
```diff
@@ -511,8 +511,13 @@ def _createDatasetSubscriptionsInDBSBuffer(self):
         existingTransaction = self.beginTransaction()
 
         subInfo = self.wmSpec.getSubscriptionInformation()
+
         for dataset in subInfo:
-            dbsDataset = DBSBufferDataset(path=dataset)
```
Comment (amaltaro) [inline src/python/WMCore/WorkQueue/WMBSHelper.py]: Are you sure this is the correct place for fetching and setting this parentage info? To me it seems it will only add IF the workflow was assigned with data subscriptions set.

**Review by amaltaro — CHANGES_REQUESTED** [review]
General idea looks Ok, but can you explain in plain english the logic in the first comment of this PR? I got lost a couple of times.

There are a few things to change, I made several comments along the code.
